src/dmd/root/response.d

```diff
@@ -14,200 +14,349 @@
+                    else
+                        numSlashes = 0;
+                    buffer[d++] = c;
+                    break;
                 }
-                break;
             }
-        Lnextarg:
         }
-    L2:
+    Lnextarg:
     }
-    if (recurse)
+    return recurse;
+}
+
+unittest
+{
+    auto args = Strings(4);
+    args[0] = "arg0";
+    args[1] = "arg1";
+    args[2] = "arg2";
+
+    char[] testData = "".dup;
+    size_t index = 1;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    assert(index == 1);
+
+    testData = (`\\\\"foo bar" \\\\foo \\\"foo \"foo "\"" # @comment`~'\0').dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    assert(index == 6);
+
+    assert(args[1][0 .. 9] == `\\foo bar`);
+    assert(args[2][0 .. 7] == `\\\\foo`);
+    assert(args[3][0 .. 5] == `\"foo`);
+    assert(args[4][0 .. 4] == `"foo`);
+    assert(args[5][0 .. 1] == `"`);
+
+    index = 7;
+    testData = "\t@recurse # comment\r\ntab\t\"@recurse\"\x1A after end\0".dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == true);
+    assert(index == 10);
+    assert(args[7][0 .. 8] == "@recurse");
+    assert(args[8][0 .. 3] == "tab");
+    assert(args[9][0 .. 8] == "@recurse");
+}
+
+unittest
+{
+    auto args = Strings(0);
+
+    char[] testData = "\x1A".dup;
+    size_t index = 0;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    assert(index == 0);
+
+    testData = "@\r".dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == true);
+    assert(index == 1);
+    assert(args[0][0 .. 2] == "@\0");
+
+    testData = "ä&#\0".dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    assert(index == 2);
+    assert(args[1][0 .. 5] == "ä&#\0");
+
+    testData = "one@\"word \0".dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    args[0] = "one@\"word";
+}
+
+/*********************************
+ * Try to resolve the null-terminated string cp to a null-terminated char[].
+ *
+ * The name is first searched for in the environment. If it is not
+ * there, it is searched for as a file name.
+ *
+ * Params:
+ *     cp = null-terminated string to look resolve
+ *
+ * Returns:
+ *     a mutable, manually allocated array containing the contents of the environment
+ *     variable or file, ending with a null-terminator.
+ *     The null-terminator is inside the bounds of the array.
+ *     If cp could not be resolved, null is returned.
+ */
+private char[] lookupInEnvironment(scope const(char)* cp) nothrow {
+
+    import core.stdc.stdlib: getenv;
+    import core.stdc.string: strlen;
+    import dmd.root.rmem: mem;
+
+    if (auto p = getenv(cp))
     {
-        /* Recursively expand @filename   */
-        if (response_expand(args))
-            goto noexpand;
+        char* buffer = mem.xstrdup(p);
+        if (!buffer)
+            return null;
```

Comment:
This check is redundant: https://github.com/dlang/dmd/blob/b2522da8566783491648bc104a29b42dc2dc569e/src/dmd/root/rmem.d#L39